trafficshape/listener_test.go

```diff
@@ -0,0 +1,383 @@
+	"time"
+)
+
+func TestListenerRead(t *testing.T) {
+	t.Parallel()
+
+	l, err := net.Listen("tcp", "[::1]:0")
+	if err != nil {
+		t.Fatalf("net.Listen(): got %v, want no error", err)
+	}
+
+	tsl := NewListener(l)
+	defer tsl.Close()
+
+	if got := tsl.ReadBitrate(); got != defaultBitrate {
+		t.Errorf("tsl.ReadBitrate(): got %d, want defaultBitrate", got)
+	}
+	if got := tsl.WriteBitrate(); got != defaultBitrate {
+		t.Errorf("tsl.WriteBitrate(): got %d, want defaultBitrate", got)
+	}
+
+	tsl.SetReadBitrate(40) // 4 bytes per second
+	if got, want := tsl.ReadBitrate(), int64(40); got != want {
+		t.Errorf("tsl.ReadBitrate(): got %d, want %d", got, want)
+	}
+
+	tsl.SetWriteBitrate(40) // 4 bytes per second
+	if got, want := tsl.WriteBitrate(), int64(40); got != want {
+		t.Errorf("tsl.WriteBitrate(): got %d, want %d", got, want)
+	}
+
+	tsl.SetLatency(time.Second)
+	if got, want := tsl.Latency(), time.Second; got != want {
+		t.Errorf("tsl.Latency(): got %s, want %s", got, want)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	want := bytes.Repeat([]byte("*"), 16)
+
+	go func() {
+		// Dial the local listener.
+		c, err := net.Dial("tcp", tsl.Addr().String())
+		if err != nil {
+			t.Fatalf("net.Dial(): got %v, want no error", err)
+		}
+		defer c.Close()
+
+		// Wait for the signal that it's okay to write to the connection; ensure
+		// the test is ready to read it.
+		wg.Wait()
+
+		c.Write(want)
+	}()
+
+	tsc, err := tsl.Accept()
+	if err != nil {
+		t.Fatalf("tsl.Accept(): got %v, want no error", err)
+	}
+	defer tsc.Close()
+
+	// Signal to the write goroutine that it may begin writing.
+	wg.Done()
+
+	start := time.Now()
+	got, err := ioutil.ReadAll(tsc)
+	end := time.Now()
+
+	if err != nil {
+		t.Fatalf("tsc.Read(): got %v, want no error", err)
+	}
+	if !bytes.Equal(got, want) {
+		t.Errorf("tsc.Read(): got %q, want %q", got, want)
+	}
+
+	// Breakdown of ~3s expectation:
+	// 1 second for the initial latency
+	// ~2-3 seconds for throttled read
+	//   - 4 bytes per second with 16 bytes  total = 3 seconds (first four bytes
+	//     are read immediately at the zeroth second; 0:4, 1:8, 2:12, 3:16)
+	//   - the drain ticker begins before the initial start time so some of that
+	//     tick time is unaccounted for in the difference; potentially up to a
+	//     full second (the drain interval). For example, if the ticker is 300ms
+	//     into its tick before start is calculated we will believe that the
+	//     throttled read will have occurred in 2.7s. Allow for up to drain
+	//     interval in skew to account for this and ensure the test does not
+	//     flake.
+	//
+	// The test runtime should be negligable compared the latency simulation, so
+	// we assume the ~3s (> 2.95s) is accounted for by throttling and latency in
+	// the worst case (we read and a new tick happens immediately).
+	if got, want := end.Sub(start), 2*time.Second+950*time.Millisecond; got < want {
+		t.Errorf("tsc.Read(): took %s, want at least %s", got, want)
+	}
```

Comment:
These assertions should be a range, rather than a minimum. It should probably also fail if it takes longer than 4.5ms (3 full seconds for throttle, 1 full second for latency, plus some wiggle room for the test run itself (500ms).